src/lustre/lustreParser.mly

```diff
@@ -574,25 +578,59 @@ main_annot:
+  | AT ; timestep = NUMERAL
+    { A.Reachable (Some (At (int_of_string (HString.string_of_hstring timestep)))) }
+  | FROM ; timestep = NUMERAL
+    { A.Reachable (Some (From (int_of_string (HString.string_of_hstring timestep)))) }
+  | FROM ; timestep = NUMERAL ; 
+    WITHIN ; timestep2 = NUMERAL 
+    { A.Reachable (Some (FromWithin (int_of_string (HString.string_of_hstring timestep), 
+                                     int_of_string (HString.string_of_hstring timestep2)))) }
+  | {A.Reachable None} 
+
+
 property:
+  (* Invariant properties *)
   | PROPERTY_ANNOT ; name = option(STRING) ; e = qexpr ; SEMICOLON
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
   | PROPERTY_PSBLOCKSTART ; name = option(STRING);
     e = qexpr; SEMICOLON ; PSBLOCKEND
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
   | PROPERTY_PSBLOCKSTART ; name = option(STRING);
     COLON; e = qexpr; SEMICOLON ; PSBLOCKEND
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
   | PROPERTY_SSBLOCKSTART ; name = option(STRING);
     e = qexpr ; SEMICOLON; SSBLOCKEND
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
   | PROPERTY_SSBLOCKSTART ; name = option(STRING);
     COLON; e = qexpr ; SEMICOLON; SSBLOCKEND
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
+
+  (* Reachability queries *)
+  | PROPERTY_ANNOT ; REACHABLE ; name = option(STRING) ; e = qexpr ; bound = property_timestep; SEMICOLON
+    { A.AnnotProperty (mk_pos $startpos, name, A.UnaryOp (mk_pos $startpos, A.Not, e), bound) }
+  | PROPERTY_PSBLOCKSTART ; REACHABLE ; name = option(STRING);
+    e = qexpr;  bound = property_timestep; SEMICOLON ; PSBLOCKEND
+    { A.AnnotProperty (mk_pos $startpos, name, e, bound) }
+  | PROPERTY_PSBLOCKSTART ; REACHABLE ; name = option(STRING);
+    COLON; e = qexpr; bound = property_timestep; SEMICOLON ; PSBLOCKEND
+    { A.AnnotProperty (mk_pos $startpos, name, e, bound) }
+  | PROPERTY_SSBLOCKSTART ; REACHABLE ; name = option(STRING);
+    e = qexpr ; bound = property_timestep; SEMICOLON; SSBLOCKEND
+    { A.AnnotProperty (mk_pos $startpos, name, e, bound) }
+  | PROPERTY_SSBLOCKSTART ; REACHABLE ; name = option(STRING);
+    COLON; e = qexpr ;  bound = property_timestep; SEMICOLON; SSBLOCKEND
+    { A.AnnotProperty (mk_pos $startpos, name, e, bound) }
 
 check:
   | CHECK ; name = option(STRING) ; e = qexpr ; SEMICOLON
-    { A.AnnotProperty (mk_pos $startpos, name, e) }
+    { A.AnnotProperty (mk_pos $startpos, name, e, Invariant) }
+  | CHECK ; REACHABLE ; name = option(STRING) ; e = qexpr ; bound = property_timestep; SEMICOLON
+    { A.AnnotProperty (mk_pos $startpos, name, A.UnaryOp (mk_pos $startpos, A.Not, e), bound) }
```

Comment:
For all these reachability checks, I would store the original expression in the Ast, and convert it to the negation later on.